File: src/command_dispatcher.cpp
#include "hyper_vision_agent/command_dispatcher.hpp"
#include <iostream>
#include <sstream>

namespace hyper_vision_agent {

namespace {

std::string EscapeJsonString(const std::string& input) {
    std::ostringstream ss;
    for (char c : input) {
        switch (c) {
            case '"': ss << "\\\""; break;
            case '\\': ss << "\\\\"; break;
            case '\b': ss << "\\b"; break;
            case '\f': ss << "\\f"; break;
            case '\n': ss << "\\n"; break;
            case '\r': ss << "\\r"; break;
            case '\t': ss << "\\t"; break;
            default:
                if (static_cast<unsigned char>(c) < 0x20) {
                    char buf[8];
                    snprintf(buf, sizeof(buf), "\\u%04x", c);
                    ss << buf;
                } else {
                    ss << c;
                }
        }
    }
    return ss.str();
}

} // namespace

CommandDispatcher::CommandDispatcher(std::shared_ptr<CdpConnection> connection)
    : connection_(connection) {}

bool CommandDispatcher::EnablePage(const SessionId& session_id) {
    return connection_->SendCommand("Page.enable", "{}", session_id).success;
}

bool CommandDispatcher::EnableDOM(const SessionId& session_id) {
    return connection_->SendCommand("DOM.enable", "{}", session_id).success;
}

bool CommandDispatcher::EnableRuntime(const SessionId& session_id) {
    return connection_->SendCommand("Runtime.enable", "{}", session_id).success;
}

bool CommandDispatcher::EnableNetwork(const SessionId& session_id) {
    return connection_->SendCommand("Network.enable", "{}", session_id).success;
}

CdpResponse CommandDispatcher::PageNavigate(const std::string& url, const SessionId& session_id) {
    std::string params = "{\"url\":\"" + EscapeJsonString(url) + "\"}";
    return connection_->SendCommand("Page.navigate", params, session_id);
}

CdpResponse CommandDispatcher::PageReload(const SessionId& session_id) {
    return connection_->SendCommand("Page.reload", "{\"ignoreCache\":true}", session_id);
}

CdpResponse CommandDispatcher::PageCaptureScreenshot(const std::string& format, int quality, const SessionId& session_id) {
    std::ostringstream params;
    params << "{\"format\":\"" << format << "\"";
    if (format == "jpeg") {
        params << ",\"quality\":" << quality;
    }
    params << ",\"fromSurface\":true,\"captureBeyondViewport\":false}";
    return connection_->SendCommand("Page.captureScreenshot", params.str(), session_id);
}

CdpResponse CommandDispatcher::RuntimeEvaluate(const std::string& expression, bool return_by_value, bool await_promise, const SessionId& session_id) {
    std::ostringstream params;
    params << "{\"expression\":\"" << EscapeJsonString(expression) << "\""
           << ",\"returnByValue\":" << (return_by_value ? "true" : "false")
           << ",\"awaitPromise\":" << (await_promise ? "true" : "false")
           << ",\"userGesture\":true}";
    return connection_->SendCommand("Runtime.evaluate", params.str(), session_id);
}

CdpResponse CommandDispatcher::DOMGetDocument(int depth, bool pierce, const SessionId& session_id) {
    std::ostringstream params;
    params << "{\"depth\":" << depth << ",\"pierce\":" << (pierce ? "true" : "false") << "}";
    return connection_->SendCommand("DOM.getDocument", params.str(), session_id);
}

CdpResponse CommandDispatcher::DOMQuerySelector(int node_id, const std::string& selector, const SessionId& session_id) {
    std::ostringstream params;
    params << "{\"nodeId\":" << node_id << ",\"selector\":\"" << EscapeJsonString(selector) << "\"}";
    return connection_->SendCommand("DOM.querySelector", params.str(), session_id);
}

CdpResponse CommandDispatcher::DOMQuerySelectorAll(int node_id, const std::string& selector, const SessionId& session_id) {
    std::ostringstream params;
    params << "{\"nodeId\":" << node_id << ",\"selector\":\"" << EscapeJsonString(selector) << "\"}";
    return connection_->SendCommand("DOM.querySelectorAll", params.str(), session_id);
}

CdpResponse CommandDispatcher::DOMGetBoxModel(int node_id, const SessionId& session_id) {
    std::ostringstream params;
    params << "{\"nodeId\":" << node_id << "}";
    return connection_->SendCommand("DOM.getBoxModel", params.str(), session_id);
}

CdpResponse CommandDispatcher::DOMFocus(int node_id, const SessionId& session_id) {
    std::ostringstream params;
    params << "{\"nodeId\":" << node_id << "}";
    return connection_->SendCommand("DOM.focus", params.str(), session_id);
}

CdpResponse CommandDispatcher::InputDispatchMouseEvent(const std::string& type, double x, double y, const std::string& button, int click_count, const SessionId& session_id) {
    std::ostringstream params;
    params << "{\"type\":\"" << type << "\""
           << ",\"x\":" << x
           << ",\"y\":" << y
           << ",\"button\":\"" << button << "\""
           << ",\"clickCount\":" << click_count << "}";
    return connection_->SendCommand("Input.dispatchMouseEvent", params.str(), session_id);
}

CdpResponse CommandDispatcher::InputDispatchKeyEvent(const std::string& type, const std::string& key, const std::string& text, int windows_virtual_key_code, const SessionId& session_id) {
    std::ostringstream params;
    params << "{\"type\":\"" << type << "\""
           << ",\"key\":\"" << EscapeJsonString(key) << "\"";
    if (!text.empty()) {
        params << ",\"text\":\"" << EscapeJsonString(text) << "\"";
    }
    if (windows_virtual_key_code > 0) {
        params << ",\"windowsVirtualKeyCode\":" << windows_virtual_key_code;
    }
    params << "}";
    return connection_->SendCommand("Input.dispatchKeyEvent", params.str(), session_id);
}

CdpResponse CommandDispatcher::InputInsertText(const std::string& text, const SessionId& session_id) {
    std::string params = "{\"text\":\"" + EscapeJsonString(text) + "\"}";
    return connection_->SendCommand("Input.insertText", params, session_id);
}

CdpResponse CommandDispatcher::EmulationSetDeviceMetrics(int width, int height, double device_scale_factor, bool mobile, const SessionId& session_id) {
    std::ostringstream params;
    params << "{\"width\":" << width
           << ",\"height\":" << height
           << ",\"deviceScaleFactor\":" << device_scale_factor
           << ",\"mobile\":" << (mobile ? "true" : "false") << "}";
    return connection_->SendCommand("Emulation.setDeviceMetricsOverride", params.str(), session_id);
}

CdpResponse CommandDispatcher::EmulationSetUserAgent(const std::string& user_agent, const SessionId& session_id) {
    std::string params = "{\"userAgent\":\"" + EscapeJsonString(user_agent) + "\"}";
    return connection_->SendCommand("Emulation.setUserAgentOverride", params, session_id);
}

CdpResponse CommandDispatcher::Dispatch(const std::string& method, const std::string& params_json, const SessionId& session_id) {
    return connection_->SendCommand(method, params_json, session_id);
}

} // namespace hyper_vision_agent

File: web-sdk/src/dispatcher/command-dispatcher.ts
/**
 * Command Dispatcher - High-Level CDP Domain Command Serializer
 * Port of command_dispatcher.cpp / command_dispatcher.hpp
 */

import type { CdpConnection } from "../connection/cdp-connection.js";
import type { BoxModel, CdpResponse, SessionId } from "../types.js";

export class CommandDispatcher {
  constructor(private connection: CdpConnection) {}

  public getConnection(): CdpConnection {
    return this.connection;
  }

  public async dispatch<T = Record<string, unknown>>(
    method: string,
    params?: Record<string, unknown>,
    sessionId?: SessionId
  ): Promise<CdpResponse<T>> {
    return this.connection.sendCommand<T>(method, params, sessionId);
  }

  // ==========================================
  // Page Domain
  // ==========================================
  public async pageEnable(sessionId?: SessionId): Promise<CdpResponse> {
    return this.dispatch("Page.enable", {}, sessionId);
  }

  public async pageNavigate(
    url: string,
    sessionId?: SessionId
  ): Promise<CdpResponse<{ frameId: string; loaderId?: string; errorText?: string }>> {
    return this.dispatch("Page.navigate", { url }, sessionId);
  }

  public async pageReload(ignoreCache = false, sessionId?: SessionId): Promise<CdpResponse> {
    return this.dispatch("Page.reload", { ignoreCache }, sessionId);
  }

  public async pageGetFrameTree(sessionId?: SessionId): Promise<CdpResponse> {
    return this.dispatch("Page.getFrameTree", {}, sessionId);
  }

  public async pageCaptureScreenshot(
    format = "png",
    quality = 80,
    clip?: { x: number; y: number; width: number; height: number; scale?: number },
    fromSurface = true,
    captureBeyondViewport = true,
    sessionId?: SessionId
  ): Promise<CdpResponse<{ data: string }>> {
    const params: Record<string, unknown> = {
      format,
      fromSurface,
      captureBeyondViewport,
    };
    if (format === "jpeg" || format === "webp") {
      params.quality = quality;
    }
    if (clip) {
      params.clip = {
        x: clip.x,
        y: clip.y,
        width: clip.width,
        height: clip.height,
        scale: clip.scale ?? 1.0,
      };
    }
    return this.dispatch<{ data: string }>("Page.captureScreenshot", params, sessionId);
  }

  public async pagePrintToPdf(sessionId?: SessionId): Promise<CdpResponse<{ data: string }>> {
    return this.dispatch<{ data: string }>("Page.printToPDF", {}, sessionId);
  }

  public async pageAddScriptToEvaluateOnNewDocument(
    source: string,
    sessionId?: SessionId
  ): Promise<CdpResponse<{ identifier: string }>> {
    return this.dispatch<{ identifier: string }>(
      "Page.addScriptToEvaluateOnNewDocument",
      { source },
      sessionId
    );
  }

  public async pageStartScreencast(
    format = "png",
    quality = 80,
    maxWidth = 1920,
    maxHeight = 1080,
    everyNthFrame = 1,
    sessionId?: SessionId
  ): Promise<CdpResponse> {
    return this.dispatch(
      "Page.startScreencast",
      {
        format,
        quality,
        maxWidth,
        maxHeight,
        everyNthFrame,
      },
      sessionId
    );
  }

  public async pageStopScreencast(sessionId?: SessionId): Promise<CdpResponse> {
    return this.dispatch("Page.stopScreencast", {}, sessionId);
  }

  public pageScreencastFrameAck(sessionId?: SessionId, frameNumber = 1): void {
    this.connection.sendCommandAsync("Page.screencastFrameAck", { sessionId: frameNumber }, sessionId);
  }

  // ==========================================
  // Runtime Domain
  // ==========================================
  public async runtimeEnable(sessionId?: SessionId): Promise<CdpResponse> {
    return this.dispatch("Runtime.enable", {}, sessionId);
  }

  public async runtimeEvaluate<T = unknown>(
    expression: string,
    returnByValue = true,
    awaitPromise = true,
    sessionId?: SessionId
  ): Promise<CdpResponse<{ result: { type: string; value?: T; description?: string }; exceptionDetails?: { text: string } }>> {
    return this.dispatch("Runtime.evaluate", {
      expression,
      returnByValue,
      awaitPromise,
    }, sessionId);
  }

  public async runtimeCallFunctionOn<T = unknown>(
    functionDeclaration: string,
    objectId: string,
    args: unknown[] = [],
    returnByValue = true,
    awaitPromise = true,
    sessionId?: SessionId
  ): Promise<CdpResponse<{ result: { type: string; value?: T; description?: string } }>> {
    return this.dispatch("Runtime.callFunctionOn", {
      functionDeclaration,
      objectId,
      arguments: args.map((v) => ({ value: v })),
      returnByValue,
      awaitPromise,
    }, sessionId);
  }

  // ==========================================
  // DOM Domain
  // ==========================================
  public async domEnable(sessionId?: SessionId): Promise<CdpResponse> {
    return this.dispatch("DOM.enable", {}, sessionId);
  }

  public async domGetDocument(
    depth = -1,
    pierce = true,
    sessionId?: SessionId
  ): Promise<CdpResponse<{ root: { nodeId: number } }>> {
    return this.dispatch("DOM.getDocument", { depth, pierce }, sessionId);
  }

  public async domQuerySelector(
    nodeId: number,
    selector: string,
    sessionId?: SessionId
  ): Promise<CdpResponse<{ nodeId: number }>> {
    return this.dispatch("DOM.querySelector", { nodeId, selector }, sessionId);
  }

  public async domQuerySelectorAll(
    nodeId: number,
    selector: string,
    sessionId?: SessionId
  ): Promise<CdpResponse<{ nodeIds: number[] }>> {
    return this.dispatch("DOM.querySelectorAll", { nodeId, selector }, sessionId);
  }

  public async domGetBoxModel(
    nodeId: number,
    sessionId?: SessionId
  ): Promise<CdpResponse<{ model: { content: number[]; padding: number[]; border: number[]; margin: number[]; width: number; height: number } }>> {
    return this.dispatch("DOM.getBoxModel", { nodeId }, sessionId);
  }

  public async domFocus(nodeId: number, sessionId?: SessionId): Promise<CdpResponse> {
    return this.dispatch("DOM.focus", { nodeId }, sessionId);
  }

  // ==========================================
  // Input Domain
  // ==========================================
  public async inputDispatchMouseEvent(
    type: "mousePressed" | "mouseReleased" | "mouseMoved" | "mouseWheel",
    x: number,
    y: number,
    button: "none" | "left" | "middle" | "right" = "none",
    clickCount = 0,
    modifiers = 0,
    deltaX = 0,
    deltaY = 0,
    sessionId?: SessionId
  ): Promise<CdpResponse> {
    const params: Record<string, unknown> = {
      type,
      x,
      y,
      button,
      clickCount,
      modifiers,
    };
    if (type === "mouseWheel") {
      params.deltaX = deltaX;
      params.deltaY = deltaY;
    }
    return this.dispatch("Input.dispatchMouseEvent", params, sessionId);
  }

  public async inputDispatchKeyEvent(
    type: "keyDown" | "keyUp" | "rawKeyDown" | "char",
    text?: string,
    unmodifiedText?: string,
    key?: string,
    code?: string,
    windowsVirtualKeyCode?: number,
    nativeVirtualKeyCode?: number,
    modifiers = 0,
    sessionId?: SessionId
  ): Promise<CdpResponse> {
    const params: Record<string, unknown> = {
      type,
      modifiers,
    };
    if (text !== undefined) params.text = text;
    if (unmodifiedText !== undefined) params.unmodifiedText = unmodifiedText;
    if (key !== undefined) params.key = key;
    if (code !== undefined) params.code = code;
    if (windowsVirtualKeyCode !== undefined) params.windowsVirtualKeyCode = windowsVirtualKeyCode;
    if (nativeVirtualKeyCode !== undefined) params.nativeVirtualKeyCode = nativeVirtualKeyCode;

    return this.dispatch("Input.dispatchKeyEvent", params, sessionId);
  }

  // ==========================================
  // Emulation Domain
  // ==========================================
  public async emulationSetDeviceMetricsOverride(
    width: number,
    height: number,
    deviceScaleFactor = 1.0,
    mobile = false,
    sessionId?: SessionId
  ): Promise<CdpResponse> {
    return this.dispatch(
      "Emulation.setDeviceMetricsOverride",
      {
        width,
        height,
        deviceScaleFactor,
        mobile,
      },
      sessionId
    );
  }

  public async emulationSetUserAgentOverride(
    userAgent: string,
    acceptLanguage?: string,
    platform?: string,
    sessionId?: SessionId
  ): Promise<CdpResponse> {
    const params: Record<string, unknown> = { userAgent };
    if (acceptLanguage) params.acceptLanguage = acceptLanguage;
    if (platform) params.platform = platform;
    return this.dispatch("Emulation.setUserAgentOverride", params, sessionId);
  }

  // ==========================================
  // Target Domain
  // ==========================================
  public async targetCreateTarget(
    url = "about:blank",
    width?: number,
    height?: number,
    browserContextId?: string,
    newWindow = false,
    background = false
  ): Promise<CdpResponse<{ targetId: string }>> {
    const params: Record<string, unknown> = {
      url,
      newWindow,
      background,
    };
    if (width !== undefined) params.width = width;
    if (height !== undefined) params.height = height;
    if (browserContextId) params.browserContextId = browserContextId;

    return this.dispatch<{ targetId: string }>("Target.createTarget", params);
  }

  public async targetCloseTarget(targetId: string): Promise<CdpResponse<{ success: boolean }>> {
    return this.dispatch<{ success: boolean }>("Target.closeTarget", { targetId });
  }

  public async targetAttachToTarget(
    targetId: string,
    flatten = true
  ): Promise<CdpResponse<{ sessionId: string }>> {
    return this.dispatch<{ sessionId: string }>("Target.attachToTarget", {
      targetId,
      flatten,
    });
  }

  public async targetDetachFromTarget(
    sessionId?: SessionId,
    targetId?: string
  ): Promise<CdpResponse> {
    const params: Record<string, unknown> = {};
    if (sessionId) params.sessionId = sessionId;
    if (targetId) params.targetId = targetId;
    return this.dispatch("Target.detachFromTarget", params);
  }

  public async targetGetTargets(): Promise<CdpResponse<{ targetInfos: Array<{ targetId: string; type: string; title: string; url: string; attached: boolean }> }>> {
    return this.dispatch("Target.getTargets", {});
  }
}
